envoy: optimise getWildcardNetworkPolicyRule()

jrajahalme opened the pull request:
Introduce `identity.NumericIdentitySlice` to be used instead of bare `[]identity.NumericIdentity`. This allows use of the new `NumericIdentitySlice.AsUint32Slice()` function, that returns the slice of numeric identities as a slice of `uint32`s without allocating and copying memory. This is possible since `identity.NumericIdentity` is defined as `uint32`.

Use this new `NumericIdentitySlice.AsUint32Slice()` in Envoy policy updates to make them more efficient.

Optimise `getWildcardNetworkPolicyRule()` by not using a map to dedup numeric identities, but using the new generic `SortAndDedupSlice()` to sort and deduplicate in place.

jrajahalme committed:
envoy: Fix commants with references to deprecated uint64 policy ids

Fix comments that still had references to the now-deprecated uint64
policy IDs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Spend less time in getWildcardNetworkPolicyRule

Do not use a map to dedup numeric identities, use a new generic SortAndDedupSlice()
to sort and deduplicate in place.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

nebril commented on pkg/identity/numericidentity.go:
```diff
@@ -482,6 +483,14 @@ type NumericIdentity uint32
+// AsUint32Slice returns the NumericIdentitySlice as a slice of uint32 without copying any data.
+// This is safe as long as the underlying type stays as uint32.
+func (nids NumericIdentitySlice) AsUint32Slice() []uint32 {
+	return unsafe.Slice((*uint32)(unsafe.Pointer(&nids[0])), len(nids))
```
This will panic if the slice is empty. Should we allocate new slice in that case?

jrajahalme replied:
I'll add a check and return a nil slice in if the slice is empty or nil!

jrajahalme replied:
Current use never called this with an empty slice. I adjusted use in `pkg/envoy/xds_server.go` a bit to make this clearer, too.

nebril reviewed (changes requested):
LGTM overall, one small issue found inline.

mhofstetter commented on pkg/identity/numericidentity.go:
```diff
@@ -482,6 +483,17 @@ type NumericIdentity uint32
+	if len(nids) == 0 {
+		return nil
+	}
+	return unsafe.Slice((*uint32)(unsafe.Pointer(&nids[0])), len(nids))
```
conversion to `unsafe.Pointer` can be removed?

-> `return unsafe.Slice((*uint32)(&nids[0]), len(nids))`

jrajahalme replied:
Thanks, makes this easier to understand and less prone for regressions!

mhofstetter commented on pkg/identity/numericidentity.go:
```diff
@@ -482,6 +483,17 @@ type NumericIdentity uint32
+
+// AsUint32Slice returns the NumericIdentitySlice as a slice of uint32 without copying any data.
+// This is safe as long as the underlying type stays as uint32.
+func (nids NumericIdentitySlice) AsUint32Slice() []uint32 {
```
maybe add a unit-test

jrajahalme replied:
Added!

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
+		return slice[i] < slice[j]
+	})
+
+	var n int = 1
```
type can be omitted

`var n = 1` or `n := 1`

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
+		return slice
+	}
+
+	sort.Slice(slice, func(i, j int) bool {
```
use `slices.Sort(slice)` from the new slices package

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
 
+// SortAndDedupSlice sorts the given slice and then removes duplicates from it without allocating
+// memory to do so.
+func SortAndDedupSlice[K cmp.Ordered](slice []K) []K {
```
i guess such a generic method either belongs to a different package than `envoy` or doesn't need to be exposed.

or try to get rid of it at all by directly use the suggested `slices.xxx` functions inline?

jrajahalme replied:
Will get rid of this, did not know about the new slices package!

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
+	})
+
+	var n int = 1
+	for i := 1; i < len(slice); i++ {
```
use `slices.Compact(slice)` from the new slices package to make a unique slice

-> `return slices.Compact(slice)`

mhofstetter commented on pkg/envoy/xds_server.go:
```diff
@@ -1384,6 +1385,28 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, wildcard bool, l7Parser
+// SortAndDedupSlice sorts the given slice and then removes duplicates from it without allocating
+// memory to do so.
+func SortAndDedupSlice[K cmp.Ordered](slice []K) []K {
+	if len(slice) < 2 {
```
by using `slices.xxx` functions, this additional check shouldn't be necessary.

mhofstetter reviewed (changes requested):
change by itself looks good. but added suggestions to replace some of the slice-handling stuff with new Go built-ins.

joamaki commented on pkg/envoy/xds_server.go:
```diff
@@ -1398,58 +1421,55 @@ func getWildcardNetworkPolicyRule(selectors policy.L7DataMap) *cilium.PortNetwor
+		var i int
+		for _, selections := range sels {
+			i += copy(remotePolicies[i:], selections)
+		}
```
```suggestion
		remotePolicies = make([]uint32, 0, count)
		for _, selections := range sels {
			remotePolicies = append(remotePolicies, selections...)
		}
```

https://go.dev/doc/effective_go#append

jrajahalme replied:
Thanks for this! I had assumed copy() be faster, but it turns out it is not! Changed as suggested.

mhofstetter commented on pkg/identity/numericidentity_test.go:
```diff
@@ -69,3 +70,17 @@ func TestGetAllReservedIdentities(t *testing.T) {
+		require.Equal(t, nid.Uint32(), uint32Slice[i])
+	}
+
+	slices.Sort(uint32Slice)
```
nit: the last lines are actually testing Go's slicing package. Might not really the right place to test this here.

What you actually want to do is adding an additional test for `xdsServer.getPortNetworkPolicyRule` that is using this code

jrajahalme replied:
Removed the lines using the slices package.

jrajahalme replied:
Added a new test for `getWildcardNetworkPolicyRule()`

aditighag reviewed (commented):
Cool! Could you confirm if it's safe for `NumericIdentitySlice` to reference the same memory as the passed `[]NumericIdentity`? Earlier, we used to copy the results, so this didn't matter. 
It's not obvious to me why this change is safe without any additional locking. 
There is a note that says the slice must not be modified - https://github.com/cilium/cilium/blob/main/pkg/policy/selectorcache.go#L34-L34. How do we ensure this contract holds true - https://go.dev/play/p/j1lRJQ4jSP2?

jrajahalme committed:
identity: Add NumericIdentitySlice type

Introduce NumericIdentitySlice to be used instead of bare
[]NumericIdentity. This allows use of the new
NumericIdentitySlice.AsUint32Slice() function, that returns the slice of
numeric identities as a slice of uint32's without allocating and copying
memory. This is possible since identity.NumericIdentity is defined as
uint32.

Use this new NumericIdentitySlice.AsUint32Slice() in Envoy policy updates
to make them more efficient.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aditighag reviewed (approved):
(I was added as a reviewer as part of the sig-policy group.)

rolinh reviewed (approved):
Not sure why I'm being pulled in for vendor changes but given there are non, lgtm in that respect :sweat_smile:
